fix(search_family): Process wrong field types in indexes for the FT.SEARCH and FT.AGGREGATE commands

BagritsevichStepan opened the pull request:
fixes dragonflydb#3986

romange commented on src/core/search/base.cc:
```diff
@@ -37,4 +39,11 @@ WrappedStrPtr::operator std::string_view() const {
 
+std::optional<double> ParseNumericField(std::string_view value) {
+  double value_as_double;
+  if (absl::SimpleAtod(value, &value_as_double) || value.empty())
```
` || value.empty()` looks like a bug

BagritsevichStepan replied:
Why?

From Redis:
```
127.0.0.1:6379> json.set j1 . '{"value":1}'
OK
127.0.0.1:6379> json.set j2 . '{"another_value":"not number"}'
OK
127.0.0.1:6379> ft.create index on json schema $.value as value numeric
OK
127.0.0.1:6379> ft.search index "*"
1) (integer) 2
2) "j2"
3) 1) "$"
   2) "{\"another_value\":\"not number\"}"
4) "j1"
5) 1) "$"
   2) "{\"value\":1}"
```

So here Redis indexes `j2`, even if it doesn't have the required field (i.e., the field value is an empty string).

BagritsevichStepan replied:
Yes, I found the edge case: `json.set j1 . '{"value":""}'`

romange commented on src/core/search/indices.cc:
```diff
@@ -71,19 +71,21 @@ absl::flat_hash_set<string> NormalizeTags(string_view taglist, bool case_sensiti
-    double num;
-    if (absl::SimpleAtod(str, &num))
-      entries_.emplace(num, id);
+    entries_.emplace(ConvertToDouble(str), id);
```
the assumption here is that all values are convertible to double? why is that?

BagritsevichStepan replied:
Before it was matched => proved that all fields are doubles

romange commented on src/core/search/indices.cc:
```diff
@@ -71,19 +71,21 @@ absl::flat_hash_set<string> NormalizeTags(string_view taglist, bool case_sensiti
 
+bool NumericIndex::Matches(DocId id, DocumentAccessor* doc, string_view field) {
+  auto strings = doc->GetStrings(field);
+  return std::all_of(strings.begin(), strings.end(),
```
so here you test that all field values are indexable? why all of them?

BagritsevichStepan replied:
Actually the result of GetStrings should be only one value

BagritsevichStepan replied:
From Redis:
```
127.0.0.1:6379> json.set j1 . '{"arr":[{"value":1},{"value":"one"}]}'
OK
127.0.0.1:6379> json.set j2 . '{"arr":[{"value":1},{"value":2}]}'
OK
127.0.0.1:6379> ft.create index on json schema $.arr[*].value AS value NUMERIC
OK
127.0.0.1:6379> ft.search index "*"
1) (integer) 1
2) "j2"
3) 1) "$"
   2) "{\"arr\":[{\"value\":1},{\"value\":2}]}"
```

BagritsevichStepan replied:
So all fields should have the same field type as index. Only then this document can be added

BagritsevichStepan replied:
I also see that we had another logic for the sort indices. We just parsed only the first value. It is a bug. But there is another problem that we are saving the sortable values

romange commented on src/core/search/search.h:
```diff
@@ -77,7 +77,7 @@ class FieldIndices {
   FieldIndices(const Schema& schema, const IndicesOptions& options, PMR_NS::memory_resource* mr);
 
-  void Add(DocId doc, DocumentAccessor* access);
+  bool Add(DocId doc, DocumentAccessor* access);
```
please add a comment what it returns

romange commented on src/core/search/sort_indices.cc:
```diff
@@ -66,18 +71,23 @@ template <typename T> PMR_NS::memory_resource* SimpleValueSortIndex<T>::GetMemRe
-double NumericSortIndex::Get(DocId id, DocumentAccessor* doc, std::string_view field) {
+std::optional<double> NumericSortIndex::Get(DocId id, DocumentAccessor* doc,
+                                            std::string_view field) {
   auto str = doc->GetStrings(field);
```
`str` is ambigous clear. it is a vector? either name it `str_vec` or provide an explicit type

romange commented on src/core/search/sort_indices.cc:
```diff
@@ -66,18 +71,23 @@ template <typename T> PMR_NS::memory_resource* SimpleValueSortIndex<T>::GetMemRe
-  return v;
+  auto value_as_double = ParseNumericField(str.front());
+  if (!value_as_double) {
+    LOG(WARNING) << "Failed to parse numeric value from field: " << field
```
why it's a warning? can it happen as a result of user input? is it a bug or a valid scenario?

BagritsevichStepan replied:
Yes, it can happen as a result of user input

BagritsevichStepan replied:
I think it would be better to remove this `LOG`

romange commented on src/core/search/sort_indices.cc:
```diff
@@ -66,18 +71,23 @@ template <typename T> PMR_NS::memory_resource* SimpleValueSortIndex<T>::GetMemRe
-PMR_NS::string StringSortIndex::Get(DocId id, DocumentAccessor* doc, std::string_view field) {
+std::optional<PMR_NS::string> StringSortIndex::Get(DocId id, DocumentAccessor* doc,
+                                                   std::string_view field) {
   auto str = doc->GetStrings(field);
```
same, lets call it `str_vec` or `str_arr`

romange commented on src/server/search/doc_accessors.cc:
```diff
@@ -38,13 +38,13 @@ string_view SdsToSafeSv(sds str) {
       LOG(DFATAL) << "Failed to convert " << value << " to double";
     }
-    return value_as_double;
+    return value_as_double.value();
```
this will throw imho if `value_as_double` is nullopt, so it's better use CHECK fail  above.

romange reviewed (commented):
I will wait for @dranikpg  to review , meanwhile some minor comments so I could understand the PR better

BagritsevichStepan commented:
@romange I've added more tests. After adding support for nullable values, a more tests will be added. Please take a look at the PR.

dranikpg reviewed (commented):
My two cents:
1. Please write next time what you changed in the PR description, so I can dive into the changes quicker
2. Maybe look into unifying Matches() and Add(), this would simplify code
3. Please check vectors or add a TODO

If we hurry for a release I can approve 🙂

BagritsevichStepan commented on src/core/search/indices.cc:
```diff
@@ -192,6 +207,10 @@ std::pair<size_t /*dim*/, VectorSimilarity> BaseVectorIndex::Info() const {
 
+bool BaseVectorIndex::Matches(DocId id, DocumentAccessor* doc, string_view field) {
+  return true;
+}
```
Yes, I added checking and tests

BagritsevichStepan commented on src/core/search/search.cc:
```diff
@@ -571,23 +571,38 @@ void FieldIndices::CreateSortIndices(PMR_NS::memory_resource* mr) {
+  auto it = lower_bound(all_ids_.begin(), all_ids_.end(), doc);
+  if (it == all_ids_.end() || *it != doc) {  // Document wasn't indexed
+    return;
+  }
```
Yes, you are right

BagritsevichStepan commented on src/core/search/base.h:
```diff
@@ -81,6 +81,13 @@ struct DocumentAccessor {
+  virtual bool Matches(DocId id, DocumentAccessor* doc, std::string_view field) = 0;
+
+  // Before adding the document make sure that Matches() returns true
   virtual void Add(DocId id, DocumentAccessor* doc, std::string_view field) = 0;
```
I understand that we could implement it as `bool Add()`. I did it this way because, by the next release, we might be able to implement `std::optional<Data> Matches()`, with `Add` taking this data as an argument. This way, we won’t need to do a rollback, and we’ll fetch the data only once. Moreover, we already have a cache in `JsonAccessor`.

BagritsevichStepan commented:
I found several new bugs and fixed them in the latest commit:
1. Bug:
```
JSON.SET j1 . '{"data":1}'
JSON.SET j2 . '{"data":"1"}'
FT.CREATE i1 ON JSON SCHEMA $.data AS data NUMERIC
FT.CREATE i2 ON JSON SCHEMA $.data AS data TEXT
```
`FT.SEARCH i1 "*"` and `FT.SEARCH i2 "*"` both return j1 and j2. However, it should return j1 for i1 and j2 for i2. This happens because we are parsing string and numeric fields in the same way in the `DocumentAccessor`.

2. Bug related to how we parse `FtVector` in the `GetVector` function in `DocumentAccessor`.

I’ve also added more tests

BagritsevichStepan commented:
Found and fixed another bug:

We should support queries like this:
```
JSON.SET j1 . '{"data":["first value", "second_value"]}'
FT.CREATE index ON JSON SCHEMA $.data AS data TEXT
```
`FT.SEARCH index "*"` should return `j1`

romange commented on src/core/search/base.h:
```diff
@@ -81,6 +86,13 @@ struct DocumentAccessor {
+  /* Returns true if the field type in the document matches the index and, therefore, the document
+  can be added.
+  TODO: Return the data needed for the Add() function to avoid retrieving the same data twice.*/
+  virtual bool IsValidFieldType(DocumentAccessor* doc, std::string_view field) = 0;
```
nit: please use `const Type&` wherever possible.

romange commented on src/core/search/base.h:
```diff
@@ -91,4 +103,20 @@ struct BaseSortIndex : BaseIndex {
   virtual std::vector<ResultScore> Sort(std::vector<DocId>* ids, size_t limit, bool desc) const = 0;
 };
 
+/* Used for converting field values to double. Returns false if the conversion fails */
```
```suggestion
/* Used for converting field values to double. Returns nullopt if the conversion fails */
```

romange commented on src/server/search/doc_accessors.h:
```diff
@@ -37,6 +38,33 @@ struct BaseAccessor : public search::DocumentAccessor {
   */
   virtual SearchDocData SerializeDocument(const search::Schema& schema) const;
+
+  virtual AccessResult<VectorInfo> GetVector(std::string_view active_field) const {
```
please move the implementation to cc file

dranikpg replied:
Oh, yes, and this

dranikpg commented on src/core/search/base.h:
```diff
@@ -66,13 +66,18 @@ using SortableValue = std::variant<std::monostate, double, std::string>;
 
 // Interface for accessing document values with different data structures underneath.
 struct DocumentAccessor {
+  // nullopt if field cannot be converted to type T
+  template <typename T> using AccessResult = std::optional<T>;
+
```
not shorter than std::optional and confuses the reader with a new type 😅

dranikpg reviewed (dismissed):
Good job fixing bugs around vector parsing, it wasn't safe before. 

Still, once we merge this for 1.25, I'd suggest to unite IsValidFieldType and Add, because we have safe/unsafe parsing functions, optionals that are dereferenced "by implicit contract" and double effort indexing documents (we parse it two times)
